File: implementation/sipamato/src/test/java/ch/difty/sipamato/persistance/jooq/user/UserUpdateSetStepSetterTest.java
package ch.difty.sipamato.persistance.jooq.user;

import static ch.difty.sipamato.db.tables.SipamatoUser.SIPAMATO_USER;
import static ch.difty.sipamato.persistance.jooq.RecordMapperTest.CREATED;
import static ch.difty.sipamato.persistance.jooq.RecordMapperTest.CREATED_BY;
import static ch.difty.sipamato.persistance.jooq.RecordMapperTest.LAST_MOD;
import static ch.difty.sipamato.persistance.jooq.RecordMapperTest.LAST_MOD_BY;
import static ch.difty.sipamato.persistance.jooq.RecordMapperTest.VERSION;
import static ch.difty.sipamato.persistance.jooq.user.UserRecordMapperTest.EMAIL;
import static ch.difty.sipamato.persistance.jooq.user.UserRecordMapperTest.ENABLED;
import static ch.difty.sipamato.persistance.jooq.user.UserRecordMapperTest.FIRST_NAME;
import static ch.difty.sipamato.persistance.jooq.user.UserRecordMapperTest.ID;
import static ch.difty.sipamato.persistance.jooq.user.UserRecordMapperTest.LAST_NAME;
import static ch.difty.sipamato.persistance.jooq.user.UserRecordMapperTest.PASSWORD;
import static ch.difty.sipamato.persistance.jooq.user.UserRecordMapperTest.USER_NAME;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

import org.mockito.Mock;

import ch.difty.sipamato.db.tables.records.SipamatoUserRecord;
import ch.difty.sipamato.entity.User;
import ch.difty.sipamato.persistance.jooq.UpdateSetStepSetter;
import ch.difty.sipamato.persistance.jooq.UpdateSetStepSetterTest;

public class UserUpdateSetStepSetterTest extends UpdateSetStepSetterTest<SipamatoUserRecord, User> {

    private final UpdateSetStepSetter<SipamatoUserRecord, User> setter = new UserUpdateSetStepSetter();

    @Mock
    private User entityMock;

    @Override
    protected UpdateSetStepSetter<SipamatoUserRecord, User> getSetter() {
        return setter;
    }

    @Override
    protected User getEntity() {
        return entityMock;
    }

    @Override
    protected void specificTearDown() {
        verifyNoMoreInteractions(entityMock);
    }

    @Override
    protected void entityFixture() {
        when(entityMock.getId()).thenReturn(ID);
        UserRecordMapperTest.entityFixtureWithoutIdFields(entityMock);
    }

    @Override
    protected void stepSetFixtureExceptAudit() {
        when(getStep().set(SIPAMATO_USER.USER_NAME, USER_NAME)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.FIRST_NAME, FIRST_NAME)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.LAST_NAME, LAST_NAME)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.EMAIL, EMAIL)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.PASSWORD, PASSWORD)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.ENABLED, ENABLED)).thenReturn(getMoreStep());
    }

    @Override
    protected void stepSetFixtureAudit() {
        when(getMoreStep().set(SIPAMATO_USER.CREATED, CREATED)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.CREATED_BY, CREATED_BY)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.LAST_MODIFIED, LAST_MOD)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.LAST_MODIFIED_BY, LAST_MOD_BY)).thenReturn(getMoreStep());
        when(getMoreStep().set(SIPAMATO_USER.VERSION, VERSION)).thenReturn(getMoreStep());
    }

    @Override
    protected void verifyCallToAllFieldsExceptAudit() {
        verify(entityMock).getUserName();
        verify(entityMock).getFirstName();
        verify(entityMock).getLastName();
        verify(entityMock).getEmail();
        verify(entityMock).getPassword();
        verify(entityMock).isEnabled();
    }

    @Override
    protected void verifyStepSettingExceptAudit() {
        verify(getStep()).set(SIPAMATO_USER.USER_NAME, USER_NAME);
        verify(getMoreStep()).set(SIPAMATO_USER.FIRST_NAME, FIRST_NAME);
        verify(getMoreStep()).set(SIPAMATO_USER.LAST_NAME, LAST_NAME);
        verify(getMoreStep()).set(SIPAMATO_USER.EMAIL, EMAIL);
        verify(getMoreStep()).set(SIPAMATO_USER.PASSWORD, PASSWORD);
        verify(getMoreStep()).set(SIPAMATO_USER.ENABLED, ENABLED);
    }

    @Override
    protected void verifyStepSettingAudit() {
        verify(getMoreStep()).set(SIPAMATO_USER.CREATED, CREATED);
        verify(getMoreStep()).set(SIPAMATO_USER.CREATED_BY, CREATED_BY);
        verify(getMoreStep()).set(SIPAMATO_USER.LAST_MODIFIED, LAST_MOD);
        verify(getMoreStep()).set(SIPAMATO_USER.LAST_MODIFIED_BY, LAST_MOD_BY);
        verify(getMoreStep()).set(SIPAMATO_USER.VERSION, VERSION + 1);
    }

}

File: core/core-persistence-jooq/src/test/kotlin/ch/difty/scipamato/core/persistence/user/UserUpdateSetStepSetterTest.kt
package ch.difty.scipamato.core.persistence.user

import ch.difty.scipamato.core.db.tables.ScipamatoUser.SCIPAMATO_USER
import ch.difty.scipamato.core.db.tables.records.ScipamatoUserRecord
import ch.difty.scipamato.core.entity.User
import ch.difty.scipamato.core.persistence.RecordMapperTest
import ch.difty.scipamato.core.persistence.UpdateSetStepSetter
import ch.difty.scipamato.core.persistence.UpdateSetStepSetterTest
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify

internal class UserUpdateSetStepSetterTest : UpdateSetStepSetterTest<ScipamatoUserRecord, User>() {

    override val setter: UpdateSetStepSetter<ScipamatoUserRecord, User> = UserUpdateSetStepSetter()

    override val entity = mockk<User>()

    override fun specificTearDown() {
        confirmVerified(entity)
    }

    override fun entityFixture() {
        UserRecordMapperTest.entityFixtureWithoutIdFields(entity)
    }

    override fun stepSetFixtureExceptAudit() {
        every { step.set(SCIPAMATO_USER.USER_NAME, UserRecordMapperTest.USER_NAME) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.FIRST_NAME, UserRecordMapperTest.FIRST_NAME) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.LAST_NAME, UserRecordMapperTest.LAST_NAME) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.EMAIL, UserRecordMapperTest.EMAIL) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.PASSWORD, UserRecordMapperTest.PASSWORD) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.ENABLED, UserRecordMapperTest.ENABLED) } returns moreStep
    }

    override fun stepSetFixtureAudit() {
        every { moreStep.set(SCIPAMATO_USER.CREATED, RecordMapperTest.CREATED) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.CREATED_BY, RecordMapperTest.CREATED_BY) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.LAST_MODIFIED, RecordMapperTest.LAST_MOD) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.LAST_MODIFIED_BY, RecordMapperTest.LAST_MOD_BY) } returns moreStep
        every { moreStep.set(SCIPAMATO_USER.VERSION, RecordMapperTest.VERSION + 1) } returns moreStep
    }

    override fun verifyCallToAllFieldsExceptAudit() {
        verify { entity.userName }
        verify { entity.firstName }
        verify { entity.lastName }
        verify { entity.email }
        verify { entity.password }
        verify { entity.isEnabled }
    }

    override fun verifyStepSettingExceptAudit() {
        verify { step.set(SCIPAMATO_USER.USER_NAME, UserRecordMapperTest.USER_NAME) }
        verify { moreStep.set(SCIPAMATO_USER.FIRST_NAME, UserRecordMapperTest.FIRST_NAME) }
        verify { moreStep.set(SCIPAMATO_USER.LAST_NAME, UserRecordMapperTest.LAST_NAME) }
        verify { moreStep.set(SCIPAMATO_USER.EMAIL, UserRecordMapperTest.EMAIL) }
        verify { moreStep.set(SCIPAMATO_USER.PASSWORD, UserRecordMapperTest.PASSWORD) }
        verify { moreStep.set(SCIPAMATO_USER.ENABLED, UserRecordMapperTest.ENABLED) }
    }

    override fun verifyStepSettingAudit() {
        verify { moreStep.set(SCIPAMATO_USER.CREATED, RecordMapperTest.CREATED) }
        verify { moreStep.set(SCIPAMATO_USER.CREATED_BY, RecordMapperTest.CREATED_BY) }
        verify { moreStep.set(SCIPAMATO_USER.LAST_MODIFIED, RecordMapperTest.LAST_MOD) }
        verify { moreStep.set(SCIPAMATO_USER.LAST_MODIFIED_BY, RecordMapperTest.LAST_MOD_BY) }
        verify { moreStep.set(SCIPAMATO_USER.VERSION, RecordMapperTest.VERSION + 1) }
    }
}
